### PR TITLE
[fix] Framebuffer creation issues causing black screens

### PR DESCRIPTION
**Pull request by l1ving**
Two bugs, one fix

- On Intel, creating framebuffers on Integrated GPUs which don't fully support it will cause a black screen
- On Nvidia GPUs, not creating them causes a black screen

### CODE REVIEW

File: src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt [hunk 1]
```diff
@@ -44,9 +49,18 @@ class ShaderHelper(shaderIn: ResourceLocation, vararg frameBufferNames: String)
...
 
         KamiEventBus.subscribe(this)
     }
+
+    companion object {
+        private var cachedIsIntegratedGraphics: Boolean? = null
+
+        fun isIntegratedGraphics() = cachedIsIntegratedGraphics ?: run {
+            cachedIsIntegratedGraphics = GlStateManager.glGetString(GL11.GL_VENDOR).contains("Intel")
+            cachedIsIntegratedGraphics!! // cannot be null, we just set the value
```
Comment (Xiaro) [inline src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt:63]: Just use `.apply` for this lol

File: src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt [hunk 2]
```diff
@@ -19,23 +20,27 @@ class ShaderHelper(shaderIn: ResourceLocation, vararg frameBufferNames: String)
     private val frameBufferMap = HashMap<String, Framebuffer>()
 
     init {
-        shader = if (OpenGlHelper.shadersSupported) {
+        shader = if (!OpenGlHelper.shadersSupported) {
+            KamiMod.log.warn("Shaders are unsupported by OpenGL!")
+            null
+        } else if (isIntegratedGraphics()) {
```
Comment (Xiaro) [inline src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt:26]: No `if else` chain 🗿 , use `when`

File: src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt [hunk 3]
```diff
@@ -44,9 +49,18 @@ class ShaderHelper(shaderIn: ResourceLocation, vararg frameBufferNames: String)
...
         listener<ResolutionUpdateEvent> {
-            shader?.createBindFramebuffers(it.width, it.height)
+            shader?.createBindFramebuffers(it.width, it.height) // this will not run if on Intel GPU or unsupported Shaders
         }
 
         KamiEventBus.subscribe(this)
     }
+
+    companion object {
+        private var cachedIsIntegratedGraphics: Boolean? = null
```
Comment (Xiaro) [inline src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt:59]: Why not just do this with `lateinit` and use `.isInitialized` to check

**Review by Xiaro — CHANGES_REQUESTED** [review]

**Comment by l1ving** [comment]
We don't need the mixins, @omninope confirmed it's fixed on Nvidia without them.

File: src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt (hunk 3 above)
Comment (l1ving) [reply src/main/java/me/zeroeightsix/kami/util/graphics/ShaderHelper.kt:71]: It's in a companion object

**Review by Xiaro — APPROVED** [review]

**Comment by l1ving** [comment]
Thanks. I'm waiting for 2 people to test and then I will merge

**Comment by l1ving** [comment]
Tested on Nvidia, with and without mixins it's a black screen. Apparently with mixins it's white before anything, then black once framebuffers load

**Comment by slashsynapse** [comment]
uwu

**Pull request merged**